src/irr/asset/CImageLoaderTGA.cpp

```diff
@@ -143,58 +140,69 @@ core::smart_refctd_ptr<ICPUImage> createAndconvertImageData(ICPUImage::SCreation
 
 		inputCreationImage = asset::ICPUImage::create(std::move(imgInfo));
 		inputCreationImage->setBufferAndRegions(std::move(texelBuffer), regions);
-	}
 
-	core::smart_refctd_ptr<ICPUImage> newConvertedImage;
-	{
-		auto referenceImageParams = inputCreationImage->getCreationParameters();
-		auto referenceBuffer = inputCreationImage->getBuffer();
-		auto referenceRegions = inputCreationImage->getRegions();
-		auto referenceRegion = referenceRegions.begin();
-		const auto newTexelOrBlockByteSize = asset::getTexelOrBlockBytesize(destFormat);
-
-		asset::TexelBlockInfo referenceBlockInfo(referenceImageParams.format);
-		core::vector3du32_SIMD referenceTrueExtent = referenceBlockInfo.convertTexelsToBlocks(referenceRegion->getTexelStrides());
-
-		auto newImageParams = referenceImageParams;
-		auto newCpuBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(referenceTrueExtent.X * referenceTrueExtent.Y * referenceTrueExtent.Z * newTexelOrBlockByteSize);
-		auto newRegions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<ICPUImage::SBufferCopy>>(1);
-		newRegions->front() = *referenceRegion;
-
-		newImageParams.format = destFormat;
-		newConvertedImage = ICPUImage::create(std::move(newImageParams));
-		newConvertedImage->setBufferAndRegions(std::move(newCpuBuffer), newRegions);
-	}
-
-	auto attachedRegion = newConvertedImage->getRegions().begin();
+		bool OpenGlFlip = !flip;
+		if (OpenGlFlip)
+		{
+			auto format = inputCreationImage->getCreationParameters().format;
+			asset::TexelBlockInfo blockInfo(format);
+			core::vector3du32_SIMD trueExtent = blockInfo.convertTexelsToBlocks(inputCreationImage->getRegions().begin()->getTexelStrides());
 
-	state.inImage = inputCreationImage.get();
-	state.outImage = newConvertedImage.get();
-	state.inOffset = { 0, 0, 0 };
-	state.inBaseLayer = 0;
-	state.outOffset = { 0, 0, 0 };
-	state.outBaseLayer = 0;
-	state.extent = attachedRegion->getExtent();
-	state.layerCount = attachedRegion->imageSubresource.layerCount;
-	state.inMipLevel = attachedRegion->imageSubresource.mipLevel;
-	state.outMipLevel = attachedRegion->imageSubresource.mipLevel;
+			auto copyBuffer = core::smart_refctd_ptr_static_cast<ICPUBuffer>(inputCreationImage->getBuffer()->clone());
+			auto in = reinterpret_cast<uint8_t*>(inputCreationImage->getBuffer()->getPointer());
+			auto out = reinterpret_cast<uint8_t*>(copyBuffer->getPointer()) + copyBuffer->getSize();
+			auto stride = trueExtent.X * getTexelOrBlockBytesize(format);
 
-	if (!convertFilter.execute(&state))
-		os::Printer::log("Something went wrong while converting!", ELL_WARNING);
+			for (uint32_t y = 0; y < trueExtent.Y; ++y)
+			{
+				out -= stride;
+				memcpy(in + (y * stride), out, stride);
+			}
+		}
```

Comment:
great and then you do nothing with the copy buffer :facelpalm:

Reply:
ah two memcpys

Reply:
use `swap_ranges`

Reply:
I think my code will be cleaner with that approach, but I will optimalize memcpy to get rid of it

Reply:
there is only 1 memcpy

Reply:
okay in next commit there will be no any  memcpy